model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java

```diff
@@ -0,0 +1,135 @@
+
+    private static final Logger log = Logger.getLogger(InfinispanCrlStorageProvider.class);
+
+    private final Cache<String, X509CRLEntry> crlCache;
+    private final Map<String, FutureTask<X509CRL>> tasksInProgress;
+    private final long cacheTime;
+    private final long minTimeBetweenRequests;
+
+    public InfinispanCrlStorageProvider(Cache<String, X509CRLEntry> crlCache, Map<String, FutureTask<X509CRL>> tasksInProgress,
+            int cacheTime, int minTimeBetweenRequests) {
+        this.crlCache = crlCache;
+        this.tasksInProgress = tasksInProgress;
+        this.cacheTime = cacheTime > 0? TimeUnit.SECONDS.toMillis(cacheTime) : -1;
+        this.minTimeBetweenRequests = minTimeBetweenRequests > 0? TimeUnit.SECONDS.toMillis(minTimeBetweenRequests) : 10000L;
+    }
+
+    @Override
+    public X509CRL get(String key, Callable<X509CRL> loader) throws GeneralSecurityException {
+        final X509CRLEntry crlEntry = crlCache.get(key);
+        final long currentTime = Time.currentTimeMillis();
+        if (crlEntry != null && (crlEntry.getCrl().getNextUpdate() == null || crlEntry.getCrl().getNextUpdate().compareTo(new Date(currentTime)) > 0)) {
+            log.debugf("returning CRL '%s' from cache because it's cached OK", key);
+            return crlEntry.getCrl();
+        }
+
+        // refresh the crl entry in the cache
+        return reloadCrl(key, loader, currentTime, crlEntry);
+    }
+
+    @Override
+    public boolean refreshCache(String key, Callable<X509CRL> loader) throws GeneralSecurityException {
+        final X509CRLEntry entry = crlCache.get(key);
+        final X509CRL crl = reloadCrl(key, loader, Time.currentTimeMillis(), entry);
+        return  crl != null && (entry == null || entry.getCrl() != crl);
+    }
+
+    @Override
+    public void close() {
+        // no-op
+    }
+
+    private X509CRL reloadCrl(String key, Callable<X509CRL> loader, long currentTime, X509CRLEntry crlEntry) {
+        if (crlEntry != null && currentTime < crlEntry.getLastRequestTime() + minTimeBetweenRequests){
+            log.debugf("Avoiding loading crl with key '%s' again, last refreshed time %d", key, crlEntry.getLastRequestTime());
+            return crlEntry.getCrl();
+        }
+
+        FutureTask<X509CRL> task = new FutureTask<>(() -> loadCrl(key, loader, currentTime));
+
+        final FutureTask<X509CRL> existing = tasksInProgress.putIfAbsent(key, task);
+        if (existing == null) {
+            log.debugf("Reloading crl for model key '%s'.", key);
+            task.run();
+        } else {
+            task = existing;
+        }
+
+        try {
+            return task.get();
+        } catch (ExecutionException ee) {
+            throw new RuntimeException("Error when loading crl " + key + " : " + ee.getMessage(), ee);
+        } catch (InterruptedException ie) {
+            throw new RuntimeException("Error. Interrupted when loading crl " + key, ie);
```

Comment:
restore thread interrupt signal
```suggestion
           Thread.currentThread().interrupt();
            throw new RuntimeException("Error. Interrupted when loading crl " + key, ie);
```